katpoint/projection.py

```diff
@@ -129,24 +133,196 @@
+
+
+class OutOfRangeError(ValueError):
+    """A numeric value is out of range."""
+
+
+class OutOfRange(object):
+    """Customisable handling of out-of-range numeric values."""
+
+    _treatment = _old_treatment = 'raise'
+
+    @classmethod
+    def __enter__(cls):
+        pass
+
+    @classmethod
+    def __exit__(cls, *args, **kws):
+        cls._treatment = cls._old_treatment
+
+    @classmethod
+    def get_treatment(cls):
+        """The current treatment of out-of-range values."""
+        return cls._treatment
+
+    @classmethod
+    def set_treatment(cls, treatment):
+        """"Change the treatment of out-of-range values.
+
+        The supported treatments are:
+          - 'raise': raise :class:`OutOfRangeError` (the default)
+          - 'nan': replace out-of-range values with NaNs
+          - 'clip': replace out-of-range values with nearest valid values
+
+        Parameters
+        ----------
+        treatment : {'raise', 'nan', 'clip'}
+            New treatment
+
+        Returns
+        -------
+        context_manager : :class:`OutOfRange`
+            An instance intended to be used in a with-statement
+
+        Raises
+        ------
+        ValueError
+            If `treatment` is not a recognised option
+        """
+        valid_treatments = {'raise', 'nan', 'clip'}
+        if treatment not in valid_treatments:
+            raise ValueError("Unknown out-of-range treatment '{}', must be one of {}"
+                             .format(treatment, valid_treatments))
+        cls._old_treatment = cls._treatment
+        cls._treatment = treatment
+        return cls()
+
+    @classmethod
+    def treat(cls, x, err_msg, lower=None, upper=None):
+        """Apply treatment to any out-of-range values in `x`.
+
+        Parameters
+        ----------
+        x : real number or array-like of real numbers
+            Input values (left untouched)
+        err_msg : string
+            Error message passed to exception if treatment is 'raise'
+        lower, upper : real number or None, optional
+            Bounds for values in `x` (specify at least one bound!)
+
+        Returns
+        -------
+        treated_x : float or array of float
+            Treated values (guaranteed to be in range or NaN)
+
+        Raises
+        ------
+        OutOfRangeError
+            If any values in `x` are out of range and treatment is 'raise'
+
+        Notes
+        -----
+        If a value is out of bounds by less than the machine precision, it is
+        considered a rounding error. It is not treated as out-of-range to avoid
+        false alarms, but instead silently clipped to ensure that all returned
+        data is in the valid range.
+        """
+        clipped_x = np.asarray(np.clip(x, lower, upper), dtype=np.float)
+        # Suppress false alarms due to rounding errors -> only flag substantial outliers
+        out_of_range = np.abs(x - clipped_x) > np.finfo(float).eps
```

Comment:
95% of the time `x` will be between -1 and 1 (either a cosine or somewhere within the unit circle). Nevertheless, I'll heed the first part of your request :-)

I'm keeping the dtype `float` since `clipped_x` is forced to `float` and `x - clipped_x` is then guaranteed (?) to be float thanks to coercion.

I'm also not making `rtol` customisable for now, because it will then have to flow through the entire out-of-range machinery, be better documented, etc etc and I'm tired :-D

Reply:
I've reverted to using an absolute tolerance, since I want to clip to 0 quite often and rtols are then too strict. I'm not expecting tiny numbers for `x`...